ddr/ddr/Models/GameEntities/Oncomer.swift

```diff
@@ -0,0 +1,103 @@
+
+import Foundation
+import SpriteKit
+
+class Oncomer: SKSpriteNode, Spawnable {
+  
+  let spawner: Spawner<Oncomer>
+  let emitters: [Emitter]
+  var collisionEffects: [Effect]
+  var passEffects: [Effect]
+  var goneEffects: [Effect]
+  
+  override var zPosition: CGFloat {
+    didSet {
+      // update sound
+      emitters.forEach({ $0.updateZ(zPosition) })
+      
+      // update visual
+      xScale = (GameScene.HORIZON - abs(zPosition)) / GameScene.HORIZON
+      yScale = xScale
+    }
+  }
+  
+  // override to update emitter(s)
+  override var position : CGPoint {
+    didSet {
+      emitters.forEach({ $0.updatePosition(position) })
+    }
+  }
+  
+  init (
+    spawner: Spawner<Oncomer>,
+    emitters: [Emitter],
+    collisionEffects: [Effect],
+    passEffects: [Effect],
+    goneEffects: [Effect],
+    texture: SKTexture?,
+    color: UIColor,
+    size: CGSize)
+  {
+    self.spawner = spawner
+    self.collisionEffects = collisionEffects
+    self.passEffects = passEffects
+    self.goneEffects = goneEffects
+    self.emitters = emitters
+    super.init(texture: texture, color: color, size: size)
+    self.collisionEffects = self.collisionEffects + [ KillOncomerEffect(oncomer: self) ]
+    self.passEffects = self.passEffects + [ HideEffect(nodeToHide: self) ]
+    self.goneEffects = self.goneEffects + [ KillOncomerEffect(oncomer: self) ]
+  }
+  
+  required init?(coder aDecoder: NSCoder) {
+    fatalError("init(coder:) has not been implemented")
+  }
+  
+  private func getImage() -> String {
+    return ""
+  }
+  
+  func despawn() {
+    spawner.despawn(self)
+  }
+  
+  func collidesWith(position: CGPoint) -> Bool {
```

Comment:
Why not give Oncomer the position enum instead of Bat, and handle all collision logic in Oncomer instead of being forced to override it in every subclass?

Reply:
I was not sure that we would want all oncomers to collide with the rider according to the same rules. This could be refactored if we find it useful (or maybe could be defined in a protocol extension somehow...)